### PR TITLE
[ONNX] Verification tool to pinpoint discrepancy at operator level

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88788

Example (derived from test/onnx/test_models_onnxruntime.py):
```python
import os
from typing import List, Tuple

import PIL
import torch
import torchvision
from torch.onnx import verification
from torchvision.models.detection import keypoint_rcnn

torch.manual_seed(0)


def _get_image(rel_path: str, size: Tuple[int, int]) -> torch.Tensor:
    data_dir = os.path.join(os.path.dirname(__file__), "assets")
    path = os.path.join(data_dir, *rel_path.split("/"))
    image = PIL.Image.open(path).convert("RGB").resize(size, PIL.Image.BILINEAR)

    return torchvision.transforms.ToTensor()(image)


def _get_test_images() -> Tuple[torch.Tensor, torch.Tensor]:
    return (
        _get_image("grace_hopper_517x606.jpg", (100, 320)).unsqueeze(0),
        _get_image("rgb_pytorch.png", (250, 380)).unsqueeze(0),
    )


model = keypoint_rcnn.keypointrcnn_resnet50_fpn(
    pretrained=False, pretrained_backbone=False, min_size=200, max_size=300
)
images, test_images = _get_test_images()

graph_info = verification.verify_ops(model, (images,), rtol=1e-7, atol=1e-5)

print("Received graph info: ")
assert graph_info is not None
print(graph_info.graph)
print(graph_info.input_args)
print(graph_info.params_dict)
```

Output:

```
Has mismatch:  Tensor-likes are not close!

Mismatched elements: 14 / 20 (70.0%)
Greatest absolute difference: 0.00215911865234375 at index (1, 2) (up to 1e-07 allowed)
Greatest relative difference: 0.0012423732089119571 at index (1, 0) (up to 1e-07 allowed)
-----------------------------Has mismatch in graph -----------------------------
------------------------------Check upper graph 0-------------------------------
Has mismatch:  Tensor-likes are not close!

Mismatched elements: 9822 / 10240 (95.9%)
Greatest absolute difference: 0.0002593994140625 at index (0, 42, 6, 2) (up to 1e-07 allowed)
Greatest relative difference: 0.028043646536736928 at index (0, 69, 5, 3) (up to 1e-07 allowed)
----------------------------Has mismatch in graph 0-----------------------------
------------------------------Check upper graph 00------------------------------
...
<similar logs>
...
Mismatched elements: 86048 / 655360 (13.1%)
Greatest absolute difference: 1.3113021850585938e-06 at index (0, 5, 149, 42) (up to 1e-07 allowed)
Greatest relative difference: 0.02705031426055202 at index (0, 63, 21, 5) (up to 1e-07 allowed)
------------------------Has mismatch in graph 0000110000------------------------
graph(%input.1 : Float(1, 3, 320, 128, strides=[122880, 40960, 128, 1], requires_grad=0, device=cpu),
      %backbone.body.conv1.weight : Float(64, 3, 7, 7, strides=[147, 49, 7, 1], requires_grad=1, device=cpu)):
  %aten::_convolution_1305 : NoneType = prim::Constant()
  %prim::ListConstruct_1306 : int = prim::Constant[value=2]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %prim::ListConstruct_1307 : int = prim::Constant[value=2]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %aten::_convolution_1308 : int[] = prim::ListConstruct(%prim::ListConstruct_1306, %prim::ListConstruct_1307)
  %prim::ListConstruct_1309 : int = prim::Constant[value=3]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %prim::ListConstruct_1310 : int = prim::Constant[value=3]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %aten::_convolution_1311 : int[] = prim::ListConstruct(%prim::ListConstruct_1309, %prim::ListConstruct_1310)
  %prim::ListConstruct_1312 : int = prim::Constant[value=1]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %prim::ListConstruct_1313 : int = prim::Constant[value=1]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %aten::_convolution_1314 : int[] = prim::ListConstruct(%prim::ListConstruct_1312, %prim::ListConstruct_1313)
  %aten::_convolution_1315 : bool = prim::Constant[value=0]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %prim::ListConstruct_1316 : int = prim::Constant[value=0]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %prim::ListConstruct_1317 : int = prim::Constant[value=0]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %aten::_convolution_1318 : int[] = prim::ListConstruct(%prim::ListConstruct_1316, %prim::ListConstruct_1317)
  %aten::_convolution_1319 : int = prim::Constant[value=1]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %aten::_convolution_1320 : bool = prim::Constant[value=0]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %aten::_convolution_1321 : bool = prim::Constant[value=0]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %aten::_convolution_1322 : bool = prim::Constant[value=1]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %aten::_convolution_1323 : bool = prim::Constant[value=1]() # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  %input.3 : Float(1, 64, 160, 64, strides=[655360, 10240, 64, 1], requires_grad=0, device=cpu) = aten::_convolution(%input.1, %backbone.body.conv1.weight, %aten::_convolution_1305, %aten::_convolution_1308, %aten::_convolution_1311, %aten::_convolution_1314, %aten::_convolution_1315, %aten::_convolution_1318, %aten::_convolution_1319, %aten::_convolution_1320, %aten::_convolution_1321, %aten::_convolution_1322, %aten::_convolution_1323) # /home/bowbao/pytorch_dev/torch/nn/modules/conv.py:460:0
  return (%input.3)
```

### CODE REVIEW

File: torch/onnx/verification.py [hunk 1]
```diff
@@ -731,3 +746,567 @@ def verify(
...
+
+        Returns:
+            The index of the pivot node.
+        """
+        non_const_indices = [
+            i
+            for i, n in enumerate(self.graph.nodes())
+            if n.kind() not in self._excluded_node_kinds
+        ]
+        half_idx = len(non_const_indices) // 2 - 1
```
Comment (thiagocrepaldi) [inline torch/onnx/verification.py:946]: are you using median? if so, do you need to sort the list of ints to get the middle value?

File: torch/onnx/verification.py [hunk 2]
```diff
@@ -731,3 +746,567 @@ def verify(
...
+    input_names = []
+    training = export_options.training
+    do_constant_folding = export_options.do_constant_folding
+    opset_version = export_options.opset_version
+    if params_dict is None:
+        params_dict = {}
+
+    graph_inputs = [v for v in graph.inputs()]
+    jit_inputs = tuple([arg for arg in input_args if arg is not None])
+    weights = [params_dict[v.debugName()] for v in graph_inputs[len(jit_inputs) :]]
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/verification.py:797]: Always needed. At this point it's all torch script IR graph. Weights are not stored in graph, but instead kept as inputs.

**Comment by justinchuby** [comment]
When provided a traced model:

```
  File "/home/justinchu/dev/playground/accuracy/bert_convert_accuracy.py", line 55, in <module>
    graph_info = verification.verify_ops(traced_model, (input_torch,), rtol=1e-7, atol=1e-5)
  File "<@beartype(torch.onnx.verification.verify_ops) at 0x7f2f6f80caf0>", line 307, in verify_ops
  File "/home/justinchu/dev/pytorch/torch/onnx/verification.py", line 1283, in verify_ops
    graph, params, torch_out, module = utils._create_jit_graph(model, args)
  File "/home/justinchu/dev/pytorch/torch/onnx/utils.py", line 972, in _create_jit_graph
    graph = _C._propagate_and_assign_input_shapes(
RuntimeError: input_values.size() == param_count_list.size() INTERNAL ASSERT FAILED at "/home/justinchu/dev/pytorch/torch/csrc/jit/python/script_init.cpp":491, please report a bug to PyTorch.  input_values:131 vs param_count_list:130
```

**Comment by justinchuby** [comment]
When running on a traced module:

```
Traceback (most recent call last):
  File "/home/justinchu/dev/playground/accuracy/bert_convert_accuracy.py", line 55, in <module>
    graph_info = verification.verify_ops(traced_model, (input_ids, attention_mask, token_type_ids), rtol=1e-7, atol=1e-5)
  File "<@beartype(torch.onnx.verification.verify_ops) at 0x7fee068c9a60>", line 307, in verify_ops
  File "/home/justinchu/dev/pytorch/torch/onnx/verification.py", line 1301, in verify_ops
    graph_info.find_first_mismatch(
  File "/home/justinchu/dev/pytorch/torch/onnx/verification.py", line 1165, in find_first_mismatch
    self.has_mismatch, self.pt_outs = verify_aten_graph(
  File "<@beartype(torch.onnx.verification.verify_aten_graph) at 0x7fee068dbca0>", line 251, in verify_aten_graph
  File "/home/justinchu/dev/pytorch/torch/onnx/verification.py", line 801, in verify_aten_graph
    jit_outs = torch._C._jit_interpret_graph(graph, jit_input_and_parameters)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript, serialized code (most recent call last):
  File "code/__torch__.py", line 12, in <on-demand-func>
    token_type_ids: Tensor) -> List[Tensor]:
    model = self.model
    _0 = (model).forward(input_ids, attention_mask, token_type_ids, )
          ~~~~~~~~~~~~~~ <--- HERE
    return [_0]
class MyModule(Module):

Traceback of TorchScript, original code (most recent call last):
RuntimeError: isObject() INTERNAL ASSERT FAILED at "/home/justinchu/dev/pytorch/aten/src/ATen/core/ivalue_inl.h":1549, please report a bug to PyTorch. Expected Object but got None
```

**Comment by BowenBao** [comment]
>    jit_outs = torch._C._jit_interpret_graph(graph, jit_input_and_parameters)
> RuntimeError: The following operation failed in the TorchScript interpreter.

Somehow the graph or input is not valid for `_jit_interpret_graph`. @justinchuby Could you share the log of `graph` and `jit_input_parameters` at this point?

**Comment by BowenBao** [comment]
Replaced by #89946

**Pull request closed without merging**